### PR TITLE
PR benchmarks

### PR DESCRIPTION
**Pull request by sebastienros**
Submitting for comments. I am thinking of creating a separate file that will contain which "targets" we want to support, which will include the components to builds, and the scenarios to execute. Right now it supports `/benchmark kestrel` and `/benchmark mvc`.

I also think a comment should be created to acknowledge the benchmark is running with a link to the running workflow like the backport command is doing. 

The results look like this right now:

Arguments:
```
--config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/plaintext.benchmarks.yml --scenario plaintext --application.framework net6.0 --application.options.outputFiles ./artifacts/bin/Microsoft.AspNetCore.Server.Kestrel/release/net6.0/
```
Results:
```

| application           | base                  | head                  |         |
| --------------------- | --------------------- | --------------------- | ------- |
| CPU Usage (%)         |                    99 |                    99 |   0.00% |
| Cores usage (%)       |                   397 |                   396 |  -0.25% |
| Working Set (MB)      |                    91 |                    93 |  +2.20% |
| Private Memory (MB)   |                   303 |                   311 |  +2.64% |
| Build Time (ms)       |                23,507 |                 4,342 | -81.53% |
| Start Time (ms)       |                   293 |                   295 |  +0.68% |
| Published Size (KB)   |               108,558 |               108,558 |   0.00% |
| .NET Core SDK Version | 6.0.100-rc.1.21417.19 | 6.0.100-rc.1.21417.19 |         |


| load                   | base       | head       |         |
| ---------------------- | ---------- | ---------- | ------- |
| CPU Usage (%)          |         57 |         59 |  +3.51% |
| Cores usage (%)        |        227 |        237 |  +4.41% |
| Working Set (MB)       |         40 |         42 |  +5.00% |
| Private Memory (MB)    |        358 |        358 |   0.00% |
| Start Time (ms)        |          0 |          0 |         |
| First Request (ms)     |        167 |        225 | +34.73% |
| Requests/sec           |    783,294 |    816,098 |  +4.19% |
| Requests               | 11,824,795 | 12,321,760 |  +4.20% |
| Mean latency (ms)      |       3.14 |       3.14 |   0.00% |
| Max latency (ms)       |      50.25 |      82.87 | +64.92% |
| Bad responses          |          0 |          0 |         |
| Socket errors          |          0 |          0 |         |
| Read throughput (MB/s) |      98.60 |     102.73 |  +4.19% |
| Latency 50th (ms)      |       2.84 |       2.73 |  -3.87% |
| Latency 75th (ms)      |       4.21 |       4.05 |  -3.80% |
| Latency 90th (ms)      |       5.44 |       5.28 |  -2.94% |
| Latency 99th (ms)      |      11.29 |      12.63 | +11.87% |
```

### CODE REVIEW

**Comment by sebastienros** [comment]
How it works:

1. A comment starting with `/benchmark ` is detected, and the next part is exctracted (the target)
2. The user who created the comment is checked against the repo collaborators to prevent external users from triggering builds on malicious PRs
3. The `crank` dotnet tool is installed
4. The PR base branch is checked out, and the specified target defines which component to build, e.g. Kestrel, MVC
5. A crank benchmark is executed with these binaries and stored (base.json)
6. The PR code is checked out, merged to the target branch, and built
7. A crank benchmark is executed with these binaries and stored (head.json)
8. The `crank compare` command is invoked on the results, and added to a comment on the initial PR

No artifacts are stored, so built assets can't be served publicly.
The current target machines are on Azure and isolated in a private network.
The communication to the Azure machine is done through an Azure Relay Hybrid Connection. Its connection string needs to be stored in a GitHub's repository secret, which can be revoked or rotated.

File: .github/workflows/benchmark.yml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    - name: Check user's organization
+      uses: actions/github-script@v4
+      with:
+        script: |
+
+          const repo_owner = context.payload.repository.owner.login;
+          const repo_name = context.payload.repository.name;
+          const comment_user = context.payload.comment.user.login;
+
+          var membership = await github.repos.checkCollaborator({
```
Comment (BrennanConroy) [inline .github/workflows/benchmark.yml:47]: Shouldn't this be verifying the return value?

File: .github/workflows/benchmark.yml [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+on:
+  issue_comment:
+    types: [created]
+
+jobs:
+  benchmark:
+    # comments can be triggered on issues and prs, ensure this is a PR
+    if: github.event.issue.pull_request != '' && contains(github.event.comment.body, '/benchmark')
+    # Comment out this line to test the scripts in a fork
+    # if: github.repository == 'dotnet/aspnetcore'
```
Comment (BrennanConroy) [inline .github/workflows/benchmark.yml:11]: fix

File: .github/workflows/benchmark.yml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        then
+          echo "Building Kestrel"
+          cd ./src/Servers/Kestrel
+          ./build.sh -c release
+        elif [ "${{ steps.benchmark-argument.outputs.result }}" = "mvc" ]
+        then
+          echo "Building Mvc"
+          cd ./src/Mvc
+          ./build.sh -c release -NoBuildNodeJS
+        fi
```
Comment (BrennanConroy) [inline .github/workflows/benchmark.yml:106]: else error

File: .github/workflows/benchmark.yml (hunk 1 above)
Comment (sebastienros) [reply .github/workflows/benchmark.yml:47]: It's throwing when it's wrong

**Comment by sebastienros** [comment]
Updated. I split each command in it's own step, which will allow to create custom sets of benchmarks generating custom outputs. For instance with profile matrices, or multiple scenarios, and be able to generate custom tables for instance. Since each benchmark stores the results in json, we can then use `jq` to extract which metrics to read in each run and build custom tables. The current example only show how to use `crank compare`.

File: .github/workflows/benchmark.yml [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+          const repo_owner = context.payload.repository.owner.login;
+          const repo_name = context.payload.repository.name;
+          const comment_user = context.payload.comment.user.login;
+
+          core.info(`Loading ${context.payload.issue.pull_request.url}.`);
+
+          const pull_request = (await github.request(context.payload.issue.pull_request.url)).data;
+          const target_branch = pull_request.base.ref;
+          const source_repository = pull_request.head.repo.clone_url;
+          const source_branch = pull_request.head.ref;
```
Comment (BrennanConroy) [inline .github/workflows/benchmark.yml:37]: These are all unused

File: .github/workflows/benchmark.yml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        git config --global user.email "user@company.com"
+        git merge FETCH_HEAD
+
+        echo "Building Mvc"
+        ./src/Mvc/build.sh -c release -NoBuildNodeJS
+
+        crank $CRANK_CONFIGS --profile $CRANK_PROFILE --relay AZURE_RELAY --json "$GITHUB_WORKSPACE/head.json" $CRANK_ARGUMENTS
+
+        cd $GITHUB_WORKSPACE
+        echo "Arguments:"$'\n```\n'$"$CRANK_ARGUMENTS"$'\n```\n' >> results.txt
```
Comment (BrennanConroy) [inline .github/workflows/benchmark.yml:172]: Should this bit be put in a separate step so it's used by all benchmarks?

**Review by BrennanConroy — APPROVED** [review]

File: .github/workflows/benchmark.yml (hunk 5 above)
Comment (sebastienros) [reply .github/workflows/benchmark.yml:172]: This is intentional as I expect some commands to diverge from this and call crank multiple times, or generate an output that is different. So just repeating to simplify authoring new commands.

**Comment by sebastienros** [comment]
@wtgodbe I could use a quick look from you in case you see things that are not permitted, or other concerns with github actions.

File: .github/workflows/benchmark.yml [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                repo: context.repo.repo,
+                body: message
+              });
+
+            throw message;
+          }
+
+          core.info(`Verified ${comment_user} is a repo collaborator.`);
+
+          // Verify this action is and PR comment
```
Comment (wtgodbe) [inline .github/workflows/benchmark.yml:54]: "is and" -> "is a"

**Review by wtgodbe — CHANGES_REQUESTED** [review]
This looks good to me in terms of what it's doing, but it'd be good to add a doc listing how to trigger it

File: .github/workflows/benchmark.yml (hunk 6 above)
Comment (sebastienros) [reply .github/workflows/benchmark.yml:54]:
```suggestion
          // Verify this action is a PR comment
```

File: docs/PRBenchmarks.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# Pull Request Benchmarks
+
+Pull requests that may impact performance should be benchmarked to ensure no regression is introduced.
+The most common way is to build the PR on your development machine and submit the binaries to an existing [scenario](https://github.com/aspnet/Benchmarks/tree/main/scenarios). To simplify this process this repository can do it automatically using GitHub Actions.
```
Comment (captainsafia) [inline docs/PRBenchmarks.md:4]:
```suggestion
The most common way is to build the PR on your development machine and submit the binaries to an existing [scenario](https://github.com/aspnet/Benchmarks/tree/main/scenarios). For simplicity, this process can be done automatically via GitHub Actions.

```

File: docs/PRBenchmarks.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# Pull Request Benchmarks
+
+Pull requests that may impact performance should be benchmarked to ensure no regression is introduced.
+The most common way is to build the PR on your development machine and submit the binaries to an existing [scenario](https://github.com/aspnet/Benchmarks/tree/main/scenarios). To simplify this process this repository can do it automatically using GitHub Actions.
+
+## Submitting a PR for benchmarking
+
+A special set of comments are recognized by this repository in order to trigger a benchmark. The comment starts with `/benchmark` and takes the name of the benchmark command as an argument. For instance, `/benchmark kestrel` will build the Kestrel components and submit a `plaintext` scenario benchmark to dedicated machines.
```
Comment (captainsafia) [inline docs/PRBenchmarks.md:8]: Where does one find the list of benchmarks?

Comment (sebastienros) [reply docs/PRBenchmarks.md:9]: Two lines bellow ;)

**Pull request closed without merging**